docs/design.md

```diff
@@ -0,0 +1,75 @@
+
+## Architecture
+
+The Spark Operator consists of:
+* a `SparkApplication` controller that watches events of creation, updates, and deletion of 
+`SparkApplication` objects and acts on the watch events,
+* a *submission runner* that runs `spark-submit` for submissions received from the controller,
+* a *Spark pod monitor* that watches for Spark pods and sends pod status updates to the controller,
+* a Spark pod [initializer](https://kubernetes.io/docs/admin/extensible-admission-controllers/#initializers) that performs initialization tasks on Spark driver and executor pods based on the annotations on the pods added by the controller,
+* and also a command-line tool named `sparkctl` for working with the operator. 
+
+The following diagram shows how different components interact and work together.
+
+![Architecture Diagram](architecture-diagram.png)
+
+Specifically, a user uses the `sparkctl` (or `kubectl`) to create a `SparkApplication` object. The `SparkApplication` controller receives the object through a watcher from the API server, creates a submission carrying the `spark-submit` arguments, and sends the submission to the *submission runner*. The submission runner submits the application to run and creates the driver pod of the application. Upon starting, the driver pod creates the executor pods. While the application is running, the *Spark pod monitor* watches the pods of the application and sends status updates of the pods back to the controller, which then updates the status of the application accordingly. 
+
+## The CRD Controller
+
+The `SparkApplication` controller, or CRD controller in short, watches events of creation, updates, and deletion of `SparkApplication` objects in any namespaces in a Kubernetes cluster, and acts on the watch events. When a new `SparkApplication` object is added (i.e., when the `AddFunc` callback function of the `ResourceEventHandlerFuncs` is called), it enqueues the object into an internal work queue, from which a worker picks it up prepares a submission and sends the submission to the submission runner, which actually submits the application to run in the Kubernetes cluster. The submission includes the list of arguments for the `spark-submit` command. The submission runner has a configurable number of workers for submitting applications to run in the cluster. When a `SparkApplication` object is deleted, the object is dequeued from the internal work queue and all the Kubernetes resources associated with the application get deleted or garbage collected.
```

Comment:
I'm still concerned that this is relying on event handlers in order to function correctly. Kubernetes controllers should be able to pick up at any point and begin converging toward the desired state.

Consider this scenario:

1. Your controller goes down for upgrade.
2. A SparkApplication object is created.
3. Your controller comes back up.

If you rely on getting the `AddFunc` callback to know that you haven't submitted a SparkApplication yet, you'll never converge toward the desired state (you'll never run `spark-submit`). Note that this scenario is just an example; many scenarios can cause you to miss events, which is why we don't rely on them (e.g. why we have informer resync period).

In order to be resilient, Kubernetes controllers are designed to be fully functional in the absence of any events. It must be possible to explain the design as:

1. List all objects of type SparkApplication.
2. For each SparkApplication that exists, list all objects that it depends on (Pods it owns, etc.).
3. Compare observed state with actual state and decide what action to take, if any.

In this case, step 3 might include something like: check if a Job exists to run the `spark-submit` command, and create the Job if needed.

The event handlers are merely an optimization on top of this pattern to improve response time relative to polling.

Reply:
Is the informer supposed to inform all existing objects when the operator starts (or restarts)? This is what I observed when restarting the controller.